src/libopensc/card-openpgp.c

```diff
@@ -497,6 +500,41 @@ pgp_init(sc_card_t *card)
 	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
 }
 
+/**
+ * Internal: get features of the card: capabilities, ...
+ */
+static void
+pgp_parse_hist_bytes(sc_card_t *card, u8 *ctlv, size_t ctlv_len)
+{
+	struct pgp_priv_data *priv = DRVDATA(card);
+	size_t offs;
+
+	for (offs = 0; offs < ctlv_len; offs++) {
```

Comment:
I think you should better use `sc_asn1_read_tag()` to cope with longer tags/values.

It's also possible to make this functionality available to all drivers, in `iso7816.c` via something like `int iso7816_process_hist_bytes(struct sc_card *card, unsigned long *caps, const unsigned char *buf, size_t buflen);`

Reply:
Using sc_asn1_read_tag() is not possible. 
hist_bytes use compact TLV where tag & length are encoded in one byte: 
- tag in the high nibble, 
- length of the value in the low nibble.

 This is different what sc_asn1_read_tag() deals with, and there is no need to deal with longer tags/values

I am not sure it is possible to have a generic hist_byte parsing function.
- According to what I read in the specs, the format of the hist_bytes starting with certain prefixes is defined as compact-TLV, but the individual tags and their lengths/values are not defined.
E.g. the tag 0x73 (to be painfully correct: tag 0x70 with length 3) and its meaning are specific to OpenPGP cards only.
=> If we wanted to implement a generic hist_byte parser, this would require card-specifc call-back functions.
- In addition: the only other card driver that seems to deal with hist_bytes as compact TLVs is the PIV driver.
=> very limited use today

Please let this request not block the safety improvements in hist_file parsing for OpenPGP.

Reply:
You're right regarding the use of `sc_asn1_read_tag()`.

However, I still believe that generalizing the ISO-defined compact DOs is useful. For example, you could use it in the `init()` function of the default driver (`iso7816.c`) to detect those features. (This is currently `NULL` and typically overwritten when some card driver uses `iso_ops`.) And indeed, the default selected applet in the historical bytes (as used by the PIV driver) is also ISO-defined.

Reply:
What do you mean by " ISO-defined compact DOs"?

ISO 7816-4 only talks about historical bytes in the ATR, some ISO supported formats (COMPACT-TLV), and some defined tags in those formats.
see: http://cardwerk.com/smart-card-standard-iso7816-4-section-8-historical-bytes/

ISO does not say anything about a hist_byte DO, as used in the OpenPGP card 2+ spec.

What about accepting this pull request and see whether some of it can be generalized later.

In any case, the parsing of hist_bytes in this pull request is formally correct (it parses compact-TLV), while the old one is plain wrong (it does linear search and triggers on the first occurence of a byte that looks like a tag value, but may be anything)

Please pull, and let's not make " best" the enemy of "better"!

Reply:
What about a little tit-for-tat?

I promise that I will implement something like sc_compacttlv_read_tag() and adapt the openpgp driver to it, and in return you merge this pull request.

You can use https://github.com/OpenSC/OpenSC/pull/1372 as an indication that I stick to the promises I gave.

Reply:
Agreed, thanks